woodoku/tests/entity/test_score_agent.py

```diff
@@ -0,0 +1,223 @@
+from typing import Iterable, Tuple
+
+import pytest
+from woodoku.entity.score_agent import GROUP_POINTS, STREAK_POINTS, ScoreAgent
+
+# Scoring Rules:
+#   1. 1 point for each block placed
+#   2. A single groups scores 18 points
+#   3. Each streak adds 10 points on top of the previous streak
+#   4. Each additional combo adds 28 points
+
+
+# pytest setup
+# tuples are (No. blocks placed, No. groups completed)
+@pytest.fixture(name="scorekeeper")
+def fixture_scorekeeper() -> ScoreAgent:
+    return ScoreAgent()
+
+
+# Question... (just wanna know if there is a better way)
+# Is there any way to use Mock or MagicMock to set the private state instead of
+# a hack like this or accessing the private variable
+# for example: I want to set up so that before testing the scorekeeper has a
+# score of 42 and current streak of 1
+@pytest.fixture(name="setup_and_apply_winnings", autouse=True)
+def fixture_setup_and_apply_winnings(
+    init_score: int,
+    init_streak: int,
+    winnings: Iterable[Tuple[int, int]],
+    scorekeeper: ScoreAgent,
+) -> None:
+    """
+    Automatically setup up scorekeeper and apply the winning to scorekeeper
+    for any test with with fixture/parameters
+
+    Args:
+        init_score (int): score to setup
+        init_streak (int): streak to setup
+        winnings (Iterable[Tuple[int, int]]): Contains a ordered list of winning
+            information (block placed, groups completed) to calculate score
+        scorekeeper (ScoreAgent): a new scorekeeper instance for this test
+    """
+    # setup the real world initial state if needed
+    score_before_streak = (
+        init_score
+        # remove score added by blocks (init_streak * 1)
+        - init_streak
+        # remove score added by groups (init_streak * GROUP_POINTS)
+        - init_streak * GROUP_POINTS
+        # remove scores added by streaks (0 + 10 + 20... for a total init_streak
+        # number of times i.e. [(n-1) * n]/2)
+        - init_streak * (init_streak - 1) // 2 * STREAK_POINTS
+    )
```

Comment:
Could you please explain a little more about the reason to do those substractions? I didn't follow.

Reply:
It's a little too complicated to explain if you can't make out the meaning from the code. The goal is to set up the score and the streak to the desired inputs, and as long as the assertion passes the code should be fine.

There is a better way to set up using mock that I have not implemented yet...

Reply:
Yep, may consider use mocking instead.

Reply:
I've done a little bit more investigation. I think mocking here isn't the best approach either... What I intended to mock was the get function so that I can control the output of the getters within the `calculate_winning` functions. But this then means that the test setup is dependent on the implementation of `calculate_winning`. Because what I really want to set up is the private variables and not some functions, Mocking here is also just a hack... 

Mocking as a white box approach here seems inferior to a black box approach, where I can ignore how `calculate_winning` is implemented and use it to set up the tests assuming it's correct based solely on its inputs and what it does (the scoring rules). And if the setup fails it means that `calculate winning` was not implemented correctly. which is also a test in itself.

Sticking with the current setup abstracts the implementation of `calculate winning` ... the only downside is there is a little bit of math involved to reserve engineer what scores need to be added before setting up streaks (since streak also have addition point)